qutip/core/qobj.py

```diff
@@ -1524,18 +1527,26 @@ def eigenstates(self, sparse=False, sort='low', eigvals=0,
             new_dims = [self.dims[0], [1]]
         else:
             new_dims = [self.dims[0], [1]*len(self.dims[0])]
-        ekets = np.empty((evecs.shape[1],), dtype=object)
-        ekets[:] = [Qobj(vec, dims=new_dims, copy=False)
-                    for vec in _data.split_columns(evecs, False)]
-        norms = np.array([ket.norm() for ket in ekets])
-        if phase_fix is None:
-            phase = np.array([1] * len(ekets))
-        else:
-            phase = np.array([np.abs(ket[phase_fix, 0]) / ket[phase_fix, 0]
-                              if ket[phase_fix, 0] else 1
-                              for ket in ekets])
-        return evals, ekets / norms * phase
 
+        if output_type == 'kets':
+            ekets = np.empty((evecs.shape[1],), dtype=object)
+            ekets[:] = [Qobj(vec, dims=new_dims, copy=False)
+                        for vec in _data.split_columns(evecs, False)]
+            norms = np.array([ket.norm() for ket in ekets])
+            if phase_fix is None:
+                phase = np.array([1] * len(ekets))
+            else:
+                phase = np.array([np.abs(ket[phase_fix, 0]) / ket[phase_fix, 0]
+                                if ket[phase_fix, 0] else 1
+                                for ket in ekets])
+
+            return evals, ekets / norms * phase
+        elif output_type == 'operator':
+            ekets = Qobj(evecs)
+            norm = ekets.norm()
```

Comment:
Closer but still not there.
Dimensions are not set.

The output should be the same as the kets concatenated in a single matrices:
```
op = qutip.rand_herm(N)
_, kets = op.eigenstates(output_type='kets', phase_fix=M)
_, oper = op.eigenstates(output_type='operator', phase_fix=M)
qutip.Qobj(np.hstack([vec.full() for vec in kets]), dims) == oper
```
The norm of a matrix, per default it the trace norm: `tr(sqrt(A @ A.adjoint()))`, while for the ket it's l2 norm: `sqrt(sum(|a_i|**2))`. 

For normalization and phase_fix, you may want to take a look into `Qobj.inv`, `qutip.qdiags`.